Birthday-2024-Project/Scripts/GridLogic.gd

```diff
@@ -1,89 +1,146 @@
-			gridSpaceInfo.currentStatus = GridSpaceInfo.GridSpaceStatus.CLOSED
-			gridSpaceInfo.occupyingPiece = null
-			_gridSpaces[x][y] = gridSpaceInfo
-	while availablePieces.size() > 0:
-		var piece : Node2D = availablePieces.pop_back()
-		piece.queue_free()
-	while blockerPieces.size() > 0:
-		var piece : Node2D = blockerPieces.pop_back()
-		piece.queue_free()
+	global_position = gridCenterOffset
+
+func ImportLevel(jsonImportData : String):
+	var levelSetup : LevelSetup = LevelSetup.new()
+	levelSetup.jsonData = jsonImportData
+	LoadLevel(levelSetup)
+	
+func ReloadLevel():
+	if _levelData == null:
+		ClearLevel()
+	else:
+		LoadLevel(_levelData)
 
 func LoadLevel(levelSetupData : LevelSetup):
 	ClearLevel()
+	_levelData = levelSetupData
 	
 	var pieceSetupsData : Array[PieceSetup] = levelSetupData.RetrieveLevelData()
 	if pieceSetupsData.size() > 0:
 		for pieceSetup in pieceSetupsData:
-			#look up scene piece via ID
-			var scenePiecePrefab = load("res://ScenePrefabs/Pieces/" + pieceSetup.pieceID + ".tscn")
-			#instantiate scene piece
-			var scenePiece = scenePiecePrefab.instantiate()
-			add_child(scenePiece)
-			var pieceLogic : PieceLogic = scenePiece as PieceLogic
-			pieceLogic.levelGridReference = self
-			pieceLogic.game_manager = gameManager
-			pieceLogic._initialize()
+			var pieceLogic : PieceLogic = LoadPiece(pieceSetup.pieceID)
 			pieceLogic.SetPieceRotation(pieceSetup.pieceRotation)
-			if pieceSetup.isBlocker:
+			pieceLogic.RecordGridPosition(pieceSetup.gridPosition)
+			if pieceLogic.isBlocker:
 				#a locked piece that sits on the grid
 				SetGridSpacesByPieceShape(pieceSetup.gridPosition, GridSpaceInfo.GridSpaceStatus.CLOSED, pieceLogic)
-				blockerPieces.push_back(scenePiece)
+				pieceLogic.return_piece(true)
 			else:
 				#open level grid spaces
 				SetGridSpacesByPieceShape(pieceSetup.gridPosition, GridSpaceInfo.GridSpaceStatus.OPEN, pieceLogic)
-				availablePieces.push_back(scenePiece)
 		
 		#center grid map based on the width and height
 		var midPoint : Vector2i = _GridCoordinateToPosition(Vector2i(xMaxGrid, yMaxGrid)) + _GridCoordinateToPosition(Vector2i(xMinGrid, yMinGrid))
 		global_position = midPoint * -0.5 + gridCenterOffset
 		
-		#set piece to outskirts of level grid and unrotate
-		#TODO: place more neatly around the grid
-		var iter = 0
-		for availablePiece in availablePieces:
-			availablePiece.global_position = GridCoordinateToPosition(startingPositions[iter]) #shunting it to the right side
-			(availablePiece as PieceLogic).SetPieceRotation(PieceLogic.RotationStates.DEG_0)
-			(availablePiece as PieceLogic)._SetReturnPoint()
-			iter += 1
+		if(gridMode == GridMode.PLAY):
+			#set piece to outskirts of level grid and unrotate
+			#TODO: place more neatly around the grid
+			var iter = 0
+			for availablePiece in availablePieces:
+				availablePiece.global_position = GridCoordinateToPosition(startingPositions[iter]) #shunting it to the right side
```

Comment:
having more than 8 pieces will crash the game; referencing startingPositions at a nonexistent index

Reply:
This is related to issue https://github.com/Saplings-Projects/Birthday-2024/issues/3

The 8 starting locations are fixed and not final. Just temporary for testing levels

Reply:
Positions being fixed in other PR is okay but I'm very iffy about merging changes that uncover state that leads to crash. Would it be okay to increase this number at least as a temporary solution?

Reply:
I'll do you one better. After doing enough grid math, I figured out a decent way to spawn them randomly on the sides for issue https://github.com/Saplings-Projects/Birthday-2024/issues/3
Now it can accommodate theoretically infinite number of pieces